azure/scope/machine.go

```diff
@@ -168,8 +182,16 @@ func (m *MachineScope) NICSpecs() []azure.NICSpec {
 			spec.PublicLBAddressPoolName = m.APIServerLBPoolName(m.APIServerLBName())
 		}
 	}
-	// If Nat Gateway is not specified, then the NIC needs to reference the LB to get outbound traffic.
-	if m.Role() == infrav1.Node && m.NodeNatGateway().Name == "" {
+
+	var isNatGatewayEnabled bool
+	for _, subnet := range m.Subnets() {
+		if subnet.Name == subnetName && subnet.NatGateway.Name != "" {
+			isNatGatewayEnabled = true
+		}
+	}
+
+	// If Nat Gateway is not enabled, then the NIC needs to reference the LB to get outbound traffic.
```

Comment:
> This seems a bit conflicting with the earlier logic to not set node outbound lb defaults. In a scenario where, one subnet has ngw, and one does not, we'll end up in situation where node outbound lb default wont be set, but we'll be referencing it here. Is my understanding correct?

Wait no. In that scenario, node outbound LB defaults are set. If only one subnet has not defined the nat gateway, we fallback to the outbound LB to make sure outbound traffic always works.

Reply:
Ah right, got it. But in the above logic, `isNatGatewayEnabled ` will be `true` even if one of the subnets doesn't have nat gateway, right? shouldn't we check for the opposite, as in all the subnets don't have natgw; or something similar to `oneSubnetWithoutNatGateway`?

Reply:
For some reason I'm getting caught up with the asymmetry in logic of deciding to fallback to outbound lb. My thinking was whatever logic was used earlier to set outbound lb should be used here as well.

Reply:
On a longer term though, should we try to get rid of outbound lb and just have nat gw for outbound connections? cc: @CecileRobertMichon

Reply:
possibly, Azure docs rate it as the "best" option for outbound connections https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections#scenarios

definitely not in this PR though

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1532

Reply:
> For some reason I'm getting caught up with the asymmetry in logic of deciding to fallback to outbound lb. My thinking was whatever logic was used earlier to set outbound lb should be used here as well.

It's a different case than above because we don't need to check the nat gateway for all subnets to see if one of the subnets doesn't have nat gateway (`oneSubnetWithoutNatGateway`). In this context, we are in the `MachineScope`, that means that we know exactly which subnet this machine is using and we can directly check if the used subnet has nat gateway. If it's not enabled, then we fallback to the outbound LB.

At the end, it's the same logic. If there is one subnet without nat gateway / if this machine's subnet is without nat gateway, then we fallback to the outbound LB.

Does that make sense now?

Reply:
yep, make sense. Thanks for the detailed explanation. Is it possible to abstract the details in `machineScope.Subnet()`, and perhaps add a method  `Subnet.IsNatGatewayEnabled()` so that we can do something like `m.Subnet().IsNatGatewayEnabled()`

Reply:
I'm not sure I understood the last comment, but I gave it a try nonetheless :smile: 
I added this helper method and used it wherever I could
```
// IsNatGatewayEnabled returns whether or not a Nat Gateway is enabled on the subnet.
func (s SubnetSpec) IsNatGatewayEnabled() bool {
	return s.NatGateway.Name != ""
}
```

I also changed `MachineScope.Subnet()` to return the actual subnet instead of the subnet's name
```
// Subnet returns the machine's subnet.
func (m *MachineScope) Subnet() infrav1.SubnetSpec {
	for _, subnet := range m.Subnets() {
		if subnet.Name == m.AzureMachine.Spec.SubnetName {
			return subnet
		}
	}

	return infrav1.SubnetSpec{}
}
```

so I no longer need to loop through the subnets from different places.

Reply:
Yeah, this is what I was suggesting. Thank you for refactoring!